File: whatsapp-templates/javascript/whatsapp-templates.js
/**
 * SendSeven API - WhatsApp Templates Example
 *
 * Demonstrates how to list and send WhatsApp template messages using the SendSeven API.
 * Features:
 * - List available templates
 * - Send template with text parameters
 * - Send template with header (image/document)
 * - Handle template categories (marketing, utility, authentication)
 * - Error handling for template not found, unapproved templates
 */

require('dotenv').config();

// Configuration from environment
const API_TOKEN = process.env.SENDSEVEN_API_TOKEN;
const TENANT_ID = process.env.SENDSEVEN_TENANT_ID;
const API_URL = process.env.SENDSEVEN_API_URL || 'https://api.sendseven.com/api/v1';
const CHANNEL_ID = process.env.CHANNEL_ID;
const CONTACT_ID = process.env.CONTACT_ID;

/**
 * Get common headers for API requests.
 */
function getHeaders() {
  return {
    'Authorization': `Bearer ${API_TOKEN}`,
    'X-Tenant-ID': TENANT_ID,
    'Content-Type': 'application/json',
  };
}

/**
 * List available WhatsApp templates.
 *
 * @param {string} [category] - Filter by category (MARKETING, UTILITY, AUTHENTICATION)
 * @param {string} [status='APPROVED'] - Filter by status
 * @returns {Promise<Array>} List of template objects
 */
async function listTemplates(category = null, status = 'APPROVED') {
  const params = new URLSearchParams({ status });
  if (category) {
    params.append('category', category);
  }

  const response = await fetch(`${API_URL}/whatsapp/templates?${params}`, {
    method: 'GET',
    headers: getHeaders(),
  });

  if (!response.ok) {
    const errorBody = await response.text();
    throw new Error(`API Error ${response.status}: ${errorBody}`);
  }

  const data = await response.json();
  return data.items || data;
}

/**
 * Send a WhatsApp template message.
 *
 * @param {string} channelId - The UUID of the WhatsApp channel
 * @param {string} contactId - The UUID of the contact to send to
 * @param {string} templateName - Name of the approved template
 * @param {string} [languageCode='en'] - Language code
 * @param {Array} [components=[]] - Template components with parameters
 * @returns {Promise<Object>} The created message object
 */
async function sendTemplateMessage(channelId, contactId, templateName, languageCode = 'en', components = []) {
  const payload = {
    channel_id: channelId,
    contact_id: contactId,
    template_name: templateName,
    language_code: languageCode,
  };

  if (components.length > 0) {
    payload.components = components;
  }

  const response = await fetch(`${API_URL}/messages/send/template`, {
    method: 'POST',
    headers: getHeaders(),
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    const errorBody = await response.text();
    throw new Error(`API Error ${response.status}: ${errorBody}`);
  }

  return response.json();
}

/**
 * Send a template message with text parameters in the body.
 *
 * @param {string} channelId - The UUID of the WhatsApp channel
 * @param {string} contactId - The UUID of the contact
 * @param {string} templateName - Name of the approved template
 * @param {Array<string>} bodyParams - List of text values for body placeholders
 * @param {string} [languageCode='en'] - Language code
 * @returns {Promise<Object>} The created message object
 */
async function sendTemplateWithTextParams(channelId, contactId, templateName, bodyParams, languageCode = 'en') {
  const components = [
    {
      type: 'body',
      parameters: bodyParams.map(param => ({ type: 'text', text: param })),
    },
  ];

  return sendTemplateMessage(channelId, contactId, templateName, languageCode, components);
}

/**
 * Send a template message with an image header.
 *
 * @param {string} channelId - The UUID of the WhatsApp channel
 * @param {string} contactId - The UUID of the contact
 * @param {string} templateName - Name of the approved template
 * @param {string} imageUrl - URL of the header image
 * @param {Array<string>} [bodyParams=[]] - Optional text values for body placeholders
 * @param {string} [languageCode='en'] - Language code
 * @returns {Promise<Object>} The created message object
 */
async function sendTemplateWithHeaderImage(channelId, contactId, templateName, imageUrl, bodyParams = [], languageCode = 'en') {
  const components = [
    {
      type: 'header',
      parameters: [{ type: 'image', image: { link: imageUrl } }],
    },
  ];

  if (bodyParams.length > 0) {
    components.push({
      type: 'body',
      parameters: bodyParams.map(param => ({ type: 'text', text: param })),
    });
  }

  return sendTemplateMessage(channelId, contactId, templateName, languageCode, components);
}

/**
 * Send a template message with a document header.
 *
 * @param {string} channelId - The UUID of the WhatsApp channel
 * @param {string} contactId - The UUID of the contact
 * @param {string} templateName - Name of the approved template
 * @param {string} documentUrl - URL of the document
 * @param {string} filename - Display filename for the document
 * @param {Array<string>} [bodyParams=[]] - Optional text values for body placeholders
 * @param {string} [languageCode='en'] - Language code
 * @returns {Promise<Object>} The created message object
 */
async function sendTemplateWithHeaderDocument(channelId, contactId, templateName, documentUrl, filename, bodyParams = [], languageCode = 'en') {
  const components = [
    {
      type: 'header',
      parameters: [{ type: 'document', document: { link: documentUrl, filename: filename } }],
    },
  ];

  if (bodyParams.length > 0) {
    components.push({
      type: 'body',
      parameters: bodyParams.map(param => ({ type: 'text', text: param })),
    });
  }

  return sendTemplateMessage(channelId, contactId, templateName, languageCode, components);
}

/**
 * Handle and display template-specific errors.
 *
 * @param {Error} error - The error object
 */
function handleTemplateError(error) {
  const message = error.message;
  const statusMatch = message.match(/API Error (\d+)/);
  const statusCode = statusMatch ? parseInt(statusMatch[1]) : 0;

  if (statusCode === 404) {
    console.log(`Template not found: ${message}`);
    console.log('Tip: Verify the template name exists and is approved');
  } else if (statusCode === 400) {
    if (message.toLowerCase().includes('not approved')) {
      console.log(`Template not approved: ${message}`);
      console.log('Tip: Only APPROVED templates can be sent');
    } else if (message.toLowerCase().includes('parameter')) {
      console.log(`Parameter mismatch: ${message}`);
      console.log('Tip: Ensure the number of parameters matches the template');
    } else {
      console.log(`Bad request: ${message}`);
    }
  } else if (statusCode === 401) {
    console.log('Authentication failed: Check your API token');
  } else if (statusCode === 403) {
    console.log('Permission denied: Token may lack required scopes');
  } else {
    console.log(`Error: ${message}`);
  }
}

/**
 * Validate required configuration.
 */
function validateConfig() {
  const missing = [];
  if (!API_TOKEN) missing.push('SENDSEVEN_API_TOKEN');
  if (!TENANT_ID) missing.push('SENDSEVEN_TENANT_ID');
  if (!CHANNEL_ID) missing.push('CHANNEL_ID');
  if (!CONTACT_ID) missing.push('CONTACT_ID');

  if (missing.length > 0) {
    console.error('Error: Missing required environment variables:');
    missing.forEach(v => console.error(`  - ${v}`));
    return false;
  }
  return true;
}

async function main() {
  if (!validateConfig()) {
    process.exit(1);
  }

  // Example 1: List all approved templates
  console.log('='.repeat(60));
  console.log('Listing approved WhatsApp templates...');
  console.log('='.repeat(60));

  try {
    const templates = await listTemplates();
    if (templates.length === 0) {
      console.log('No approved templates found.');
      console.log('Create templates in the WhatsApp Business Manager first.');
      return;
    }

    console.log(`Found ${templates.length} template(s):\n`);
    templates.slice(0, 5).forEach(template => {
      console.log(`  Name: ${template.name}`);
      console.log(`  Category: ${template.category}`);
      console.log(`  Language: ${template.language}`);
      console.log(`  Status: ${template.status}`);
      console.log();
    });
  } catch (error) {
    handleTemplateError(error);
    return;
  }

  // Example 2: List templates by category
  console.log('='.repeat(60));
  console.log('Listing MARKETING templates...');
  console.log('='.repeat(60));

  try {
    const marketingTemplates = await listTemplates('MARKETING');
    console.log(`Found ${marketingTemplates.length} marketing template(s)`);
  } catch (error) {
    handleTemplateError(error);
  }

  // Example 3: Send a template with text parameters
  console.log('\n' + '='.repeat(60));
  console.log('Sending template with text parameters...');
  console.log('='.repeat(60));

  try {
    const message = await sendTemplateWithTextParams(
      CHANNEL_ID,
      CONTACT_ID,
      'order_confirmation',
      ['John Doe', 'ORD-12345'],
      'en'
    );

    console.log('Template message sent successfully!');
    console.log(`  Message ID: ${message.id}`);
    console.log(`  Status: ${message.status}`);
  } catch (error) {
    handleTemplateError(error);
    console.log('\nNote: Update template_name to match your approved template');
  }

  // Example 4: Send template with image header
  console.log('\n' + '='.repeat(60));
  console.log('Sending template with image header...');
  console.log('='.repeat(60));

  try {
    const message = await sendTemplateWithHeaderImage(
      CHANNEL_ID,
      CONTACT_ID,
      'promotion_with_image',
      'https://example.com/promo-image.jpg',
      ['Summer Sale', '50%'],
      'en'
    );

    console.log('Template with image sent successfully!');
    console.log(`  Message ID: ${message.id}`);
  } catch (error) {
    handleTemplateError(error);
    console.log('\nNote: Update template_name to match your approved template');
  }

  // Example 5: Send template with document header
  console.log('\n' + '='.repeat(60));
  console.log('Sending template with document header...');
  console.log('='.repeat(60));

  try {
    const message = await sendTemplateWithHeaderDocument(
      CHANNEL_ID,
      CONTACT_ID,
      'invoice_template',
      'https://example.com/invoice.pdf',
      'Invoice-2026-001.pdf',
      ['$199.99'],
      'en'
    );

    console.log('Template with document sent successfully!');
    console.log(`  Message ID: ${message.id}`);
  } catch (error) {
    handleTemplateError(error);
    console.log('\nNote: Update template_name to match your approved template');
  }
}

main();

File: whatsapp-templates/python/whatsapp_templates.py
#!/usr/bin/env python3
"""
SendSeven API - WhatsApp Templates Example

Demonstrates how to list and send WhatsApp template messages using the SendSeven API.
Features:
- List available templates
- Send template with text parameters
- Send template with header (image/document)
- Handle template categories (marketing, utility, authentication)
- Error handling for template not found, unapproved templates
"""

import os
import requests
from dotenv import load_dotenv
from typing import Optional

# Load environment variables from .env file
load_dotenv()

# Configuration from environment
API_TOKEN = os.getenv("SENDSEVEN_API_TOKEN")
TENANT_ID = os.getenv("SENDSEVEN_TENANT_ID")
API_URL = os.getenv("SENDSEVEN_API_URL", "https://api.sendseven.com/api/v1")
CHANNEL_ID = os.getenv("CHANNEL_ID")
CONTACT_ID = os.getenv("CONTACT_ID")


def get_headers() -> dict:
    """Get common headers for API requests."""
    return {
        "Authorization": f"Bearer {API_TOKEN}",
        "X-Tenant-ID": TENANT_ID,
        "Content-Type": "application/json",
    }


def list_templates(category: Optional[str] = None, status: str = "APPROVED") -> list:
    """
    List available WhatsApp templates.

    Args:
        category: Filter by category (MARKETING, UTILITY, AUTHENTICATION)
        status: Filter by status (default: APPROVED)

    Returns:
        list: List of template objects

    Raises:
        requests.HTTPError: If the API request fails
    """
    url = f"{API_URL}/whatsapp/templates"
    params = {"status": status}
    if category:
        params["category"] = category

    response = requests.get(url, headers=get_headers(), params=params)
    response.raise_for_status()

    data = response.json()
    return data.get("items", data) if isinstance(data, dict) else data


def send_template_message(
    channel_id: str,
    contact_id: str,
    template_name: str,
    language_code: str = "en",
    components: Optional[list] = None,
) -> dict:
    """
    Send a WhatsApp template message.

    Args:
        channel_id: The UUID of the WhatsApp channel
        contact_id: The UUID of the contact to send to
        template_name: Name of the approved template
        language_code: Language code (default: en)
        components: Template components with parameters

    Returns:
        dict: The created message object

    Raises:
        requests.HTTPError: If the API request fails
    """
    url = f"{API_URL}/messages/send/template"

    payload = {
        "channel_id": channel_id,
        "contact_id": contact_id,
        "template_name": template_name,
        "language_code": language_code,
    }

    if components:
        payload["components"] = components

    response = requests.post(url, json=payload, headers=get_headers())
    response.raise_for_status()

    return response.json()


def send_template_with_text_params(
    channel_id: str,
    contact_id: str,
    template_name: str,
    body_params: list[str],
    language_code: str = "en",
) -> dict:
    """
    Send a template message with text parameters in the body.

    Args:
        channel_id: The UUID of the WhatsApp channel
        contact_id: The UUID of the contact
        template_name: Name of the approved template
        body_params: List of text values for body placeholders
        language_code: Language code (default: en)

    Returns:
        dict: The created message object
    """
    components = [
        {
            "type": "body",
            "parameters": [{"type": "text", "text": param} for param in body_params],
        }
    ]

    return send_template_message(
        channel_id=channel_id,
        contact_id=contact_id,
        template_name=template_name,
        language_code=language_code,
        components=components,
    )


def send_template_with_header_image(
    channel_id: str,
    contact_id: str,
    template_name: str,
    image_url: str,
    body_params: Optional[list[str]] = None,
    language_code: str = "en",
) -> dict:
    """
    Send a template message with an image header.

    Args:
        channel_id: The UUID of the WhatsApp channel
        contact_id: The UUID of the contact
        template_name: Name of the approved template
        image_url: URL of the header image
        body_params: Optional list of text values for body placeholders
        language_code: Language code (default: en)

    Returns:
        dict: The created message object
    """
    components = [
        {
            "type": "header",
            "parameters": [{"type": "image", "image": {"link": image_url}}],
        }
    ]

    if body_params:
        components.append(
            {
                "type": "body",
                "parameters": [{"type": "text", "text": param} for param in body_params],
            }
        )

    return send_template_message(
        channel_id=channel_id,
        contact_id=contact_id,
        template_name=template_name,
        language_code=language_code,
        components=components,
    )


def send_template_with_header_document(
    channel_id: str,
    contact_id: str,
    template_name: str,
    document_url: str,
    filename: str,
    body_params: Optional[list[str]] = None,
    language_code: str = "en",
) -> dict:
    """
    Send a template message with a document header.

    Args:
        channel_id: The UUID of the WhatsApp channel
        contact_id: The UUID of the contact
        template_name: Name of the approved template
        document_url: URL of the document
        filename: Display filename for the document
        body_params: Optional list of text values for body placeholders
        language_code: Language code (default: en)

    Returns:
        dict: The created message object
    """
    components = [
        {
            "type": "header",
            "parameters": [
                {
                    "type": "document",
                    "document": {"link": document_url, "filename": filename},
                }
            ],
        }
    ]

    if body_params:
        components.append(
            {
                "type": "body",
                "parameters": [{"type": "text", "text": param} for param in body_params],
            }
        )

    return send_template_message(
        channel_id=channel_id,
        contact_id=contact_id,
        template_name=template_name,
        language_code=language_code,
        components=components,
    )


def handle_template_error(error: requests.HTTPError) -> None:
    """Handle and display template-specific errors."""
    status_code = error.response.status_code
    try:
        error_body = error.response.json()
        error_message = error_body.get("detail", error.response.text)
    except Exception:
        error_message = error.response.text

    if status_code == 404:
        print(f"Template not found: {error_message}")
        print("Tip: Verify the template name exists and is approved")
    elif status_code == 400:
        if "not approved" in error_message.lower():
            print(f"Template not approved: {error_message}")
            print("Tip: Only APPROVED templates can be sent")
        elif "parameter" in error_message.lower():
            print(f"Parameter mismatch: {error_message}")
            print("Tip: Ensure the number of parameters matches the template")
        else:
            print(f"Bad request: {error_message}")
    elif status_code == 401:
        print("Authentication failed: Check your API token")
    elif status_code == 403:
        print("Permission denied: Token may lack required scopes")
    else:
        print(f"API Error {status_code}: {error_message}")


def validate_config() -> bool:
    """Validate required configuration."""
    missing = []
    if not API_TOKEN:
        missing.append("SENDSEVEN_API_TOKEN")
    if not TENANT_ID:
        missing.append("SENDSEVEN_TENANT_ID")
    if not CHANNEL_ID:
        missing.append("CHANNEL_ID")
    if not CONTACT_ID:
        missing.append("CONTACT_ID")

    if missing:
        print("Error: Missing required environment variables:")
        for var in missing:
            print(f"  - {var}")
        return False
    return True


def main():
    if not validate_config():
        return

    # Example 1: List all approved templates
    print("=" * 60)
    print("Listing approved WhatsApp templates...")
    print("=" * 60)

    try:
        templates = list_templates()
        if not templates:
            print("No approved templates found.")
            print("Create templates in the WhatsApp Business Manager first.")
            return

        print(f"Found {len(templates)} template(s):\n")
        for template in templates[:5]:  # Show first 5
            print(f"  Name: {template.get('name')}")
            print(f"  Category: {template.get('category')}")
            print(f"  Language: {template.get('language')}")
            print(f"  Status: {template.get('status')}")
            print()

    except requests.HTTPError as e:
        handle_template_error(e)
        return

    # Example 2: List templates by category
    print("=" * 60)
    print("Listing MARKETING templates...")
    print("=" * 60)

    try:
        marketing_templates = list_templates(category="MARKETING")
        print(f"Found {len(marketing_templates)} marketing template(s)")
    except requests.HTTPError as e:
        handle_template_error(e)

    # Example 3: Send a template with text parameters
    print("\n" + "=" * 60)
    print("Sending template with text parameters...")
    print("=" * 60)

    try:
        # Example: order_confirmation template with customer name and order ID
        message = send_template_with_text_params(
            channel_id=CHANNEL_ID,
            contact_id=CONTACT_ID,
            template_name="order_confirmation",
            body_params=["John Doe", "ORD-12345"],
            language_code="en",
        )

        print("Template message sent successfully!")
        print(f"  Message ID: {message.get('id')}")
        print(f"  Status: {message.get('status')}")

    except requests.HTTPError as e:
        handle_template_error(e)
        print("\nNote: Update template_name to match your approved template")

    # Example 4: Send template with image header
    print("\n" + "=" * 60)
    print("Sending template with image header...")
    print("=" * 60)

    try:
        message = send_template_with_header_image(
            channel_id=CHANNEL_ID,
            contact_id=CONTACT_ID,
            template_name="promotion_with_image",
            image_url="https://example.com/promo-image.jpg",
            body_params=["Summer Sale", "50%"],
            language_code="en",
        )

        print("Template with image sent successfully!")
        print(f"  Message ID: {message.get('id')}")

    except requests.HTTPError as e:
        handle_template_error(e)
        print("\nNote: Update template_name to match your approved template")

    # Example 5: Send template with document header
    print("\n" + "=" * 60)
    print("Sending template with document header...")
    print("=" * 60)

    try:
        message = send_template_with_header_document(
            channel_id=CHANNEL_ID,
            contact_id=CONTACT_ID,
            template_name="invoice_template",
            document_url="https://example.com/invoice.pdf",
            filename="Invoice-2026-001.pdf",
            body_params=["$199.99"],
            language_code="en",
        )

        print("Template with document sent successfully!")
        print(f"  Message ID: {message.get('id')}")

    except requests.HTTPError as e:
        handle_template_error(e)
        print("\nNote: Update template_name to match your approved template")


if __name__ == "__main__":
    main()
